core/src/test/scala/unit/kafka/server/TxnOffsetCommitRequestTest.scala

```diff
@@ -0,0 +1,133 @@
+      new ClusterConfigProperty(key = GroupCoordinatorConfig.GROUP_COORDINATOR_REBALANCE_PROTOCOLS_CONFIG, value = "classic"),
+      new ClusterConfigProperty(key = GroupCoordinatorConfig.OFFSETS_TOPIC_PARTITIONS_CONFIG, value = "1"),
+      new ClusterConfigProperty(key = GroupCoordinatorConfig.OFFSETS_TOPIC_REPLICATION_FACTOR_CONFIG, value = "1"),
+      new ClusterConfigProperty(key = TransactionLogConfig.TRANSACTIONS_TOPIC_PARTITIONS_CONFIG, value = "1"),
+      new ClusterConfigProperty(key = TransactionLogConfig.TRANSACTIONS_TOPIC_REPLICATION_FACTOR_CONFIG, value = "1"),
+  ))
+  def testTxnOffsetCommitWithOldConsumerGroupProtocolAndOldGroupCoordinator(): Unit = {
+    testTxnOffsetCommit(false)
+  }
+
+  private def testTxnOffsetCommit(useNewProtocol: Boolean): Unit = {
+    if (useNewProtocol && !isNewGroupCoordinatorEnabled) {
+      fail("Cannot use the new protocol with the old group coordinator.")
+    }
+
+    val topic = "topic"
+    val partition = 0
+    val transactionalId = "txn"
+    val groupId = "group"
+
+    // Creates the __consumer_offsets topics because it won't be created automatically
+    // in this test because it does not use FindCoordinator API.
+    createOffsetsTopic()
+
+    // Join the consumer group. Note that we don't heartbeat here so we must use
+    // a session long enough for the duration of the test.
+    joinConsumerGroup(groupId, useNewProtocol)
+
+    var adminClient: Admin = null
+    var producer: Producer[String, String] = null
+
+    try {
+      adminClient = cluster.admin()
+      adminClient.createTopics(singleton(new NewTopic(topic, 1, 1.toShort)))
+
+      producer = cluster.producer(Collections.singletonMap(ProducerConfig.TRANSACTIONAL_ID_CONFIG, transactionalId))
+
+      producer.initTransactions()
+      producer.beginTransaction()
+      // add this line to avoid error INVALID_TXN_STATE
+      producer.sendOffsetsToTransaction(
+        Collections.singletonMap(new TopicPartition("topic", 0), new OffsetAndMetadata(5)),
+        new ConsumerGroupMetadata(groupId))
+
+      // verify that the TXN_OFFSET_COMMIT request is processed correctly when member id is UNKNOWN_MEMBER_ID
+      // and generation id is UNKNOWN_GENERATION_ID under all api versions
+      for (version <- 0 to ApiKeys.TXN_OFFSET_COMMIT.latestVersion(isUnstableApiEnabled)) {
+        commitTxnOffset(
+          groupId = groupId,
+          memberId = JoinGroupRequest.UNKNOWN_MEMBER_ID,
+          generationId = JoinGroupRequest.UNKNOWN_GENERATION_ID,
+          producerId = 0,
+          producerEpoch = 0.toShort,
+          transactionalId = transactionalId,
+          topic = topic,
+          partition = partition,
+          offset = 100L,
+          expectedError = Errors.NONE,
+          version = version.toShort,
+        )
+      }
+    } finally {
+      adminClient.close()
+      producer.close()
+    }
+  }
+}
```

Comment:
While we are here, do you mind adding a test which verifies that the fields are correctly validated when provided?